crates/sui-core/src/authority_active/checkpoint_driver/mod.rs

```diff
@@ -0,0 +1,878 @@
+            .lock()
+            .handle_checkpoint_certificate(&past, &None)
+        {
+            warn!("Error handling certificate: {:?}", err);
+        }
+    }
+
+    let full_sync_start = latest_checkpoint
+        .map(|chk| match chk {
+            AuthenticatedCheckpoint::Signed(signed) => signed.checkpoint.sequence_number + 1,
+            AuthenticatedCheckpoint::Certified(cert) => cert.checkpoint.sequence_number + 1,
+            AuthenticatedCheckpoint::None => unreachable!(),
+        })
+        .unwrap_or(0);
+
+    for seq in full_sync_start..latest_known_checkpoint.checkpoint.sequence_number {
+        debug!("Full Sync ({:?}): {:?}", name, seq);
+        let (past, _contents) =
+            get_one_checkpoint(net.clone(), seq, true, &available_authorities).await?;
+
+        if let Err(err) = checkpoint_db
+            .lock()
+            .handle_checkpoint_certificate(&past, &_contents)
+        {
+            warn!("Sync Err: {:?}", err);
+        }
+    }
+
+    Ok(())
+}
+
+/// Gets one checkpoint certificate and optionally its contents. Note this must be
+/// given a checkpoint number that the validator knows exists, for examples because
+/// they have seen a subsequent certificate.
+#[allow(clippy::collapsible_match)]
+pub async fn get_one_checkpoint<A>(
+    net: Arc<AuthorityAggregator<A>>,
+    sequence_number: CheckpointSequenceNumber,
+    contents: bool,
+    available_authorities: &BTreeSet<AuthorityName>,
+) -> Result<(CertifiedCheckpoint, Option<CheckpointContents>), SuiError>
+where
+    A: AuthorityAPI + Send + Sync + 'static + Clone,
+{
+    let mut available_authorities = available_authorities.clone();
+    while !available_authorities.is_empty() {
+        // Get a random authority by stake
+        let sample_authority = net.committee.sample();
+        if !available_authorities.contains(sample_authority) {
+            // We want to pick an authority that has the checkpoint and its full history.
+            continue;
+        }
+
+        available_authorities.remove(sample_authority);
+
+        // Note: safe to do lookup since authority comes from the committee sample
+        //       so this should not panic.
+        let client = net.clone_client(sample_authority);
+        match client
+            .handle_checkpoint(CheckpointRequest::past(sequence_number, contents))
+            .await
+        {
+            Ok(CheckpointResponse {
+                info: AuthorityCheckpointInfo::Past(AuthenticatedCheckpoint::Certified(past)),
+                detail,
+            }) => {
+                return Ok((past, detail));
+            }
+            Ok(resp) => {
+                warn!("Sync Error: Unexpected answer: {:?}", resp);
+            }
+            Err(err) => {
+                warn!("Sync Error: peer error: {:?}", err);
+            }
+        }
+    }
+
+    Err(SuiError::GenericAuthorityError {
+        error: "Ran out of authorities.".to_string(),
```

Comment:
Does it mean we have no honest quorum of authorities?

Reply:
clarified the message